cupyx/scipy/interpolate/_bspline.py

```diff
@@ -0,0 +1,918 @@
+
+import operator
+from itertools import product
+
+import cupy
+from cupy._core import internal
+from cupy._core._scalar import get_typename
+
+from cupyx.scipy.sparse import csr_matrix
+
+import numpy as np
+
+TYPES = ['float', 'double', 'complex<float>', 'complex<double>']
+INT_TYPES = ['int', 'long long']
+
+INTERVAL_KERNEL = r'''
+#include <cupy/complex.cuh>
+
+template<typename T>
+__global__ void find_interval(
+        const T* t, const T* x, long long* out,
+        int k, int n, bool extrapolate) {
+
+    int idx = blockDim.x * blockIdx.x + threadIdx.x;
```

Comment:
Don't we need to check the boundary for `idx` so that it doesn't make overrun?

Reply:
Also, I *think* the binary search needs to account for multiple knots. These can occur anywhere, incl at either end of the knot array. The scipy routine uses a linear search, https://github.com/scipy/scipy/blob/v1.9.3/scipy/interpolate/_bspl.pyx#L32, which is literally translated from FITPACK, https://github.com/scipy/scipy/blob/v1.9.3/scipy/interpolate/fitpack/splev.f#L93

Reply:
The linear search (after the binary one) is called once an initial interval is found: https://github.com/cupy/cupy/pull/7086/commits/9869f1842780cac21d57ba0896ea2c4fb1f182d2